src/data/nat/hyperoperation.lean

```diff
@@ -0,0 +1,127 @@
+/-
+Copyright (c) 2023 Mark Andrew Gerads. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mark Andrew Gerads, Junyan Xu, Eric Wieser
+-/
+import tactic.ring
+import data.nat.parity
+
+/-!
+# Hyperoperation sequence
+
+This file defines the Hyperoperation sequence.
+`hyperoperation 0 m k = k + 1`
+`hyperoperation 1 m k = m + k`
+`hyperoperation 2 m k = m * k`
+`hyperoperation 3 m k = m ^ k`
+`hyperoperation (n + 3) m 0 = 1`
+`hyperoperation (n + 1) m (k + 1) = hyperoperation n m (hyperoperation (n + 1) m k)`
+
+## References
+
+* <https://en.wikipedia.org/wiki/Hyperoperation>
+
+## Tags
+
+hyperoperation
+-/
+
+/--
+Implementation of the hyperoperation sequence
+where `hyperoperation n m k` is the `n`th hyperoperation between `m` and `k`.
+-/
+def hyperoperation : ℕ → ℕ → ℕ → ℕ
+| 0 _ k := k + 1
+| 1 m 0 := m
+| 2 _ 0 := 0
+| (n + 3) _ 0 := 1
+| (n + 1) m (k + 1) := hyperoperation n m (hyperoperation (n + 1) m k)
+
+-- Basic hyperoperation lemmas
+
+@[simp] lemma hyperoperation_zero (m : ℕ) : hyperoperation 0 m = nat.succ :=
+funext $ λ k, by rw [hyperoperation, nat.succ_eq_add_one]
+
+lemma hyperoperation_ge_three_eq_one (n m : ℕ) : hyperoperation (n + 3) m 0 = 1 :=
+by rw hyperoperation
+
+lemma hyperoperation_recursion (n m k : ℕ) :
+  hyperoperation (n + 1) m (k + 1) = hyperoperation n m (hyperoperation (n + 1) m k) :=
+by obtain (_|_|_) := n; rw hyperoperation
+
+-- Interesting hyperoperation lemmas
+
+@[simp] lemma hyperoperation_one : hyperoperation 1 = (+) :=
+begin
+  ext m k,
+  induction k with bn bih,
+  rw [nat_add_zero m, hyperoperation],
+  rw [hyperoperation_recursion, bih, hyperoperation_zero],
+  exact nat.add_assoc m bn 1,
+end
+
+@[simp] lemma hyperoperation_two : hyperoperation 2 = (*) :=
+begin
+  ext m k,
+  induction k with bn bih,
+  rw hyperoperation,
+  exact (nat.mul_zero m).symm,
+  rw [hyperoperation_recursion, hyperoperation_one, bih],
+  ring,
+end
+
+@[simp] lemma hyperoperation_three : hyperoperation 3 = (^) :=
+begin
+  ext m k,
+  induction k with bn bih,
```

Comment:
Here and below: after induction you have two goals so need to now follow the mathlib conventions by putting the proofs of each goal in curly braces. Just looking at your code now it's hard to see where proofs are ending.

Reply:
Okay. I actually had curly braces for multiple goals before, but removed them when the linter complained. I'll look through mathlib to see an example of how to use curly braces correctly.

Reply:
Looks like the linter isn't complaining, so I'll mark this conversation resolved.